File: 02_Top-Level-Declarations/05_Object/java/Prog.java
public final class Prog {

	public static final void useInterface(AnInterface obj) {
		obj.mi();
	}

	public static final void main() {

		// The single instance of AnObject is accessed via
		// its static field INSTANCE.
		// A property is accessed via its accessors.
		AnObject.INSTANCE.setValue(AnObject.INSTANCE.getValue() + 1);
		AnObject.INSTANCE.mao();
		
		//AnotherObject.INSTANCE.year = 1970;
		//AnotherObject.INSTANCE.maoo();

		//useInterface(AnotherObject.INSTANCE);
	}

	public static void main(String[] args) { main(); }
}

File: 02_Top-Level-Declarations/05_Object/kotlin/prog.kt

// An object declaration describes a type and
// a single instance of that type
object AnObject {
	var value = 0
	
	fun mao() { println("AnObject(value: ${value})") }
}
 
interface AnInterface {
	fun mi() : Unit;
} 

// An object can implement interfaces
object AnotherObject : AnInterface {
	var year = 2022
	
	fun maoo() { println("AnotherObject(year: ${year})") }
	
	override fun mi() { println("Interface in Object") }
}

fun useInterface(obj : AnInterface) { obj.mi() }

fun main() {

	// The name of an object is used to
	// access its single instance.

	AnObject.value++
	AnObject.mao()
	
	AnotherObject.year = 1970
	AnotherObject.maoo()

	useInterface(AnotherObject)
}
